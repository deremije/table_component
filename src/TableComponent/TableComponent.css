div.tableComponent {
  overflow-y: scroll;
  overflow-x: hidden;
  height: 100%;
}
div.tC-nullState {
  overflow: hidden;
  display: flex;
  height: 100%;
}
div.tC-nullState div {
  margin: auto;
}
div.tC-scrollable {
  width: auto;
  height: auto;
  position: relative;
  display: flex;
  justify-content: flex-start;
  color: #1A1A1A;
}
.tC-column {
  height: 100%;
  border-right: solid 1px #ddd;
  padding: 0 20px;
}
.tC-header {
  padding: 10px;
  margin-bottom: 6px;
  border-bottom: solid 2px black;
  font-weight: bold;
  font-size: 18px;
  height: 32px;
  line-height: 32px;
  cursor: pointer;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  text-align: left;
}
.tC-caret {
  border-right: solid 3px black;
  border-bottom: solid 3px black;
  height: 6px;
  width: 6px;
  display: inline-block;
  margin-left: 5px;
  margin-bottom: 2px;
  transition: all linear 200ms;
}
.tC-dataItem {
  padding: 10px;
  height: 32px;
  line-height: 32px;
  font-size: 16px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  text-align: left;
}
.tC-image {
  border-radius: 50%;
  height: 24px;
  position: relative;
  top: 5px;
  width: auto;
  margin-right: 8px;
}
