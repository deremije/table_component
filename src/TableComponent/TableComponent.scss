div.tableComponent {
  width: auto;
  height: auto;
  position: relative;
  display: flex;
  justify-content: flex-start;
  color: #1A1A1A;
  .column {
    height: auto;
    border-right: solid 1px #ccc;
    padding: 0 20px;
    .header {
      padding: 10px;
      margin-bottom: 6px;
      border-bottom: solid 2px black;
      font-weight: bold;
      font-size: 18px;
      height: 32px;
      line-height: 32px;
      cursor: pointer;
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
      text-align: left;
      .caret {
        border-right: solid 3px black;
        border-bottom: solid 3px black;
        height: 6px;
        width: 6px;
        display: inline-block;
        margin-left: 5px;
        margin-bottom: 2px;
        transition: all linear 200ms;
      }
    }
    .dataItem {
      padding: 10px;
      height: 32px;
      line-height: 32px;
      font-size: 16px;
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
      text-align: left;
    }
    .imgWrapper {
      height: 48px;
      padding: 2px;
      text-align: center;
      overflow: hidden;
      img {
        border-radius: 50%;
        height: 100%;
        width: auto;
        margin: auto;
      }
    }
  }


















  // .headerRow {
  //   width: auto;
  //   padding: 0 20px;
  //   margin: 20px;
  //   height: 20px;
  //   line-height: 20px;
  //   display: flex;
  //   justify-content: flex-start;
  //   .headerItem {
  //     width: 246px;
  //     height: 20px;
  //     line-height: 20px;
  //     padding: 10px;
  //     border: solid 2px #1B0F1C;
  //     font-weight: bold;
  //     background-color: #906FA1;
  //     color: #eee;
  //     text-align: left;

  //   }
  // }
  // .tableRow {
  //   width: auto;
  //   padding: 0 20px;
  //   margin: 20px;
  //   height: 20px;
  //   line-height: 20px;
  //   display: flex;
  //   justify-content: flex-start;
  //   div {
  //     width: 246px;
  //     height: 20px;
  //     line-height: 20px;
  //     padding: 10px;
  //     border: solid 2px #1B0F1C;
  //     font-weight: bold;
  //     color: #1a1a1a;
  //     text-align: left;
  //   }
  // }
}
