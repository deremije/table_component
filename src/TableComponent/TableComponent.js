import React from 'react'
import { Suspense, lazy } from 'react';
import { Resizable } from 're-resizable';

import './TableComponent.css'

const DataItem = lazy(() => import("./DataItem/DataItem"))

class TableComponent extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            ascending: true,
            sortBy: "",
            lines: 40,
            chunkSize: 10,
            pixelBuffer: 200,
        }
        // set ref to measure scroll height
        this.element = React.createRef()
    }
    
    // override defaults via optional props, start lazy loading of data chunks
    componentDidMount() {
        this.OverrideDefaultState()
        this.logErrors()
        setTimeout(() => { this.getDataChunk() }, 0);
    }

    OverrideDefaultState() {
        if (this.props.chunkSize && typeof this.props.chunkSize === "number") this.setState({chunkSize: this.props.chunkSize, lines: this.props.chunkSize})
        if (this.props.originalLines && typeof this.props.originalLines === "number") this.setState({lines: this.props.originalLines})
        if (this.props.pixelBuffer && typeof this.props.pixelBuffer === "number") this.setState({pixelBuffer: this.props.pixelBuffer})
        if (this.props.sortBy && typeof this.props.sortBy === "string") this.setState({sortBy: this.props.sortBy})
            else this.setState({sortBy: this.props.headers[0]})
    }

    logErrors() {
        const params = [
            { name: 'clickFunction', type: 'function' },
            { name: 'columnWidth', type: 'number' },
            { name: 'includePhoto', type: 'string' },
            { name: 'chunkSize', type: 'number' },
            { name: 'originalLines', type: 'number' },
            { name: 'pixelBuffer', type: 'number' },
            { name: 'sortBy', type: 'string' }
        ]
        params.forEach(p => {
            if (this.props[p.name] && typeof this.props[p.name] !== p.type) console.error(`TableComponent: Type error for ${p.name}`)
        })
    }

    // updates sort parameters when column header is clicked
    columnSort(header) {
        if (this.state.sortBy === header) {
            this.setState({ascending: !this.state.ascending})
        } else {
            this.setState({sortBy: header, ascending: true})
        }
    }

    // adds style to carets depending on sort parameters
    caretStyle(header) {
        let style = {
            opacity: '0',
            transform: 'rotate(45deg)'
        }
        if (header === this.state.sortBy) style.opacity = '1'
        if (!this.state.ascending) style.transform = 'rotate(-135deg)'
        return style
    }
    
    // returns array based on sort parameters in state, data prop, and lazy lines
    sortData() {
        let array = [...this.props.data]
        array.sort((a,b) => {
            if (a[this.state.sortBy] > b[this.state.sortBy]) return this.state.ascending ? 1 : -1
            if (a[this.state.sortBy] < b[this.state.sortBy]) return this.state.ascending ? -1 : 1
            return 0
        })
        return array.slice(0, this.state.lines)
    }

    scrolledToBottom() {
        return this.element.current.scrollHeight - (this.element.current.scrollTop + this.element.current.clientHeight) < this.state.pixelBuffer
    }
    // check to see if user is nearing the bottom of the scroll window, and render more lines if so
    getDataChunk() {
        if (this.dataIsLoaded() && this.scrolledToBottom()) {
            if (this.props.data.length > this.state.lines) {
                this.setState({ lines: this.state.lines + this.state.chunkSize })
                setTimeout(() => { this.getDataChunk() }, 0);
            }
        }
    }

    // determine null states
    headersAreLoaded() {return Array.isArray(this.props.headers) && this.props.headers.length > 0}
    dataIsLoaded() {return Array.isArray(this.props.data) && this.props.data.length > 0}
    
    render() {
        // check the renderable data on each render
        const sortedData = this.dataIsLoaded() ? this.sortData() : []
        
        return (
            <div className='tableComponent' ref={this.element} onScroll={() => this.getDataChunk()}>
                { /* check null states before rendering columns */
                    !this.headersAreLoaded() ? <div className='tC-nullState'><div>Waiting for Headers...</div></div> :
                    !this.dataIsLoaded() ? <div className='tC-nullState'><div>Waiting for Data...</div></div> :
                            
                    /* if headers and data arrays have both been loaded, continue */
                    <div className='tC-scrollable' >
                        {this.props.headers.map(header => 
                            <Resizable 
                                key={header}
                                defaultSize={ 
                                    this.props.columnWidth && 
                                    typeof this.props.columnWidth === "number" && 
                                    {width: this.props.columnWidth}
                                }
                                enable={{ 
                                    top:false, 
                                    right:true, 
                                    bottom:false, 
                                    left:false, 
                                    topRight:false, 
                                    bottomRight:false, 
                                    bottomLeft:false, 
                                    topLeft:false 
                                }} 
                                className='tC-column'>
                                    <div className='tC-header' onClick={() => this.columnSort(header)}>
                                        {header} <span className='tC-caret' style={this.caretStyle(header)} />
                                    </div>
                                    {sortedData.map(line => 
                                        <Suspense key={line.id} fallback={<div className='tC-dataItem'>Loading {header}...</div>}>
                                            <DataItem 
                                                line={line} 
                                                header={header} 
                                                includePhoto={this.props.includePhoto && typeof this.props.includePhoto === "string" && this.props.includePhoto} 
                                                clickFunction={this.props.clickFunction}/>
                                        </Suspense>
                                    )}
                            </Resizable>
                        )}
                    </div> 
                }
            </div>
        )
    }
}

export default TableComponent