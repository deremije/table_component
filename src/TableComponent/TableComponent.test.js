import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import TableComponent from './TableComponent'

Enzyme.configure({ adapter: new Adapter() })

describe('TableComponent', () => {
    it ('should render without crashing', () => {
        const data = [{ email: 'email', name: 'name', cell: 'cell'}]
        const headers = ['email', 'name', 'cell']
        const wrapper = shallow(<TableComponent data={data} headers={headers} />)
        expect(wrapper).toMatchSnapshot()
    })

    it('should create a header for each entry in headers array', () => {
        const data = [{ email: 'email', name: 'name', cell: 'cell'}]
        const headers = ['email', 'name', 'cell']
        const wrapper = shallow(<TableComponent data={data} headers={headers} />)
        expect(wrapper.find('.header').length).toEqual(3)
    })

    it ('should render columns for each header', () => {
        const data = [{ email: 'email', name: 'name', cell: 'cell'}]
        const headers = ['email', 'name', 'cell']
        const wrapper = shallow(<TableComponent data={data} headers={headers} />)
        expect(wrapper.find('.column').length).toEqual(3)
    })

    it ('should display text', () => {
        const data = []
        const headers = ['email', 'name', 'cell']
        const wrapper = shallow(<TableComponent data={data} headers={headers} />)
        expect(wrapper.first('.column').html()).toContain("email")
    })
})