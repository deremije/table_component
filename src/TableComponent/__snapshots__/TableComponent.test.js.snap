// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableComponent should render without crashing 1`] = `ShallowWrapper {}`;
